zebra-chain/src/transaction/unmined.rs

```diff
@@ -0,0 +1,166 @@
+//! Unmined Zcash transaction identifiers and transactions.
+//!
+//! Zebra's [`UnminedTxId`] and [`UnminedTx`] enums provide the correct ID for the
+//! transaction version. They can be used to handle transactions regardless of version,
+//! and get the [`WtxId`] or [`Hash`] when required.
+//!
+//! Transaction versions 1-4 are uniquely identified by narrow transaction IDs,
+//! so Zebra and the Zcash network protocol don't use wide transaction IDs for them.
+
+use std::sync::Arc;
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+
+use super::{
+    AuthDigest, Hash,
+    Transaction::{self, *},
+    WtxId,
+};
+
+use UnminedTxId::*;
+
+/// A unique identifier for an unmined transaction, regardless of version.
+///
+/// "The transaction ID of a version 4 or earlier transaction is the SHA-256d hash
+/// of the transaction encoding in the pre-v5 format described above.
+///
+/// The transaction ID of a version 5 transaction is as defined in [ZIP-244].
+///
+/// A v5 transaction also has a wtxid (used for example in the peer-to-peer protocol)
+/// as defined in [ZIP-239]."
+/// [Spec: Transaction Identifiers]
+///
+/// [ZIP-239]: https://zips.z.cash/zip-0239
+/// [ZIP-244]: https://zips.z.cash/zip-0244
+/// [Spec: Transaction Identifiers]: https://zips.z.cash/protocol/protocol.pdf#txnidentifiers
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub enum UnminedTxId {
+    /// A narrow unmined transaction identifier.
+    ///
+    /// Used to uniquely identify transaction versions 1-4.
+    Narrow(Hash),
+
+    /// A wide unmined transaction identifier.
+    ///
+    /// Used to uniquely identify transaction version 5.
+    Wide(WtxId),
+}
```

Comment:
These `TxId`s still apply post-mining though, right? They are not exclusively for un-mined transactions